File: 백준/Silver/2583. 영역 구하기/영역 구하기.py
from collections import deque
m,n,k=map(int,input().split())
graph=[[0]*n for _ in range(m)]
for _ in range(k):
    x1,y1,x2,y2=map(int,input().split())
    for i in range(x1,x2):
        for j in range(m-y1-1,m-y2-1,-1):
            graph[j][i]=1
ans=[]
dx=[-1,1,0,0]
dy=[0,0,-1,1]
def bfs(x,y):
    q=deque([(x,y)])
    graph[x][y]=1
    cnt=1
    while q:
        x,y=q.popleft()
        for i in range(4):
            nx,ny=x+dx[i],y+dy[i]
            if 0<=nx<m and 0<=ny<n and graph[nx][ny]==0:
                q.append((nx,ny))
                graph[nx][ny]=1
                cnt+=1
    return cnt
for i in range(m):
    for j in range(n):
        if graph[i][j]==0:
            ans.append(bfs(i,j))
ans.sort()
print(len(ans))
print(*ans)

File: 백준/Silver/2583. 영역 구하기/영역 구하기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {

    static class Point {
        int r;
        int c;

        public Point(int r, int c) {
            this.r = r;
            this.c = c;
        }
    }

    static int M, N, K;
    static int[][] map;
    static boolean[][] visited;
    static int[] dr = {-1, 1, 0, 0};
    static int[] dc = {0, 0, -1, 1};

    public static void main(String[] args) throws IOException {
        input();
        solution();
    }

    private static void input() throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        M = Integer.parseInt(st.nextToken());
        N = Integer.parseInt(st.nextToken());
        K = Integer.parseInt(st.nextToken());
        map = new int[N][M];
        for (int i = 0; i < K; i++) {
            st = new StringTokenizer(br.readLine());
            int x1 = Integer.parseInt(st.nextToken());
            int y1 = Integer.parseInt(st.nextToken());
            int x2 = Integer.parseInt(st.nextToken());
            int y2 = Integer.parseInt(st.nextToken());

            for (int r = x1; r < x2; r++) {
                for (int c = y1; c < y2; c++) {
                    map[r][c] = 1;
                }
            }
        }
    }

    private static void solution() {
        List<Integer> list = new ArrayList<>();
        visited = new boolean[N][M];
        
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if (map[i][j] == 0 && !visited[i][j]) {
                    list.add(bfs(i, j));
                }
            }
        }
        Collections.sort(list);
        System.out.println(list.size());
        for (int i = 0; i < list.size(); i++) {
            System.out.print(list.get(i) + " ");
        }
    }

    private static int bfs(int r, int c) {
        Queue<Point> q = new LinkedList<>();
        q.offer(new Point(r, c));
        visited[r][c] = true;
        int cnt = 1;

        while (!q.isEmpty()) {
            Point p = q.poll();
            r = p.r;
            c = p.c;
            for (int i = 0; i < 4; i++) {
                int nr = r + dr[i];
                int nc = c + dc[i];
                if (0 > nr || nr >= N || 0 > nc || nc >= M || visited[nr][nc] || map[nr][nc] == 1) continue;

                q.offer(new Point(nr, nc));
                visited[nr][nc] = true;
                cnt++;
            }
        }
        return cnt;
    }
}